### PR TITLE
Attaching elements

### LINKED ISSUE #807: Implement Attach for flows and stacks
Flows and Stacks should be able to attach to things: http://shoesrb.com/manual/Styles.html#:attach

I played around with this in Shoes3 and it seems like the primary purpose of this style is to fix content relative to the window (I'm not sure what good attaching a slot to another slot/element is, but that's not for me to judge! :)

For example, a fixed bottom bar (in Shoes 3 only for now):

``` ruby
Shoes.app do   
  50.times do
    stack {para "blah"}
  end

  f = flow height: 30, attach: Shoes::Window, top: app.height - 30 do
    background black
    para "Footnotes!", stroke: white
  end
end
```

### PR DESCRIPTION
**Pull request by jasonrclark**
Revises slot positioning to take advantage of attaching an element to another slot. Fixes #807.

## Sample
Sample program exercising `attach`:

```
Shoes.app do
  stack do
    para "blah"
    para "blah"
    para "blah"
    para "blah"
  end

  @s = stack do
    background blue
    para "The Body!!", stroke: white
  end

  flow height: 60, width: 50, attach: @s, left: 20, top: 30 do
    background black
    para "Footnotes!", stroke: white
  end
end
```

**Before**
image

**After**
image

## TODO

* [x] Look into `Shoes::Window` as attachment target, although that class doesn't exist at all (yet?) in Shoes 4 (http://shoesrb.com/manual/Styles.html#:attach)
* [x] Add a sample using attach... something with real purpose behind fiddling with positioning
* [x] More testing, particularly see how things work with margins
* [x] Rename `CurrentPosition` to `Position`

### CODE REVIEW

File: shoes-core/lib/shoes/slot.rb [hunk 1]
```diff
@@ -205,6 +209,15 @@ def position_element(_element, _current_position)
       raise 'position_element is a subclass responsibility'
     end
 
+    def position_attached_element(element)
+      attached_left = element.attached_to.element_left + element.style[:left].to_i
+      attached_top  = element.attached_to.element_top + element.style[:top].to_i
+
+      position_element_at element, attached_left, attached_top
+
+      CurrentPosition.new attached_left, attached_top, attached_top
```
Comment (PragTob) [inline shoes-core/lib/shoes/slot.rb]: So after we position the attached element we continue after the attached element? That is... interesting. I sort of expected it to be free floating and not influence current_position at all (like normal elements with their left/top set). Is that the behavior in Shoes 3?

**Comment by PragTob** [comment]
LGTM save for the points you lined out that still need to get done :)

File: shoes-core/lib/shoes/slot.rb [hunk 2]
```diff
@@ -183,15 +183,19 @@ def position_contents
...
       if element.needs_positioning?
-        position_element element, current_position
-        element.contents_alignment(current_position) if element.respond_to? :contents_alignment
-      end
+        if element.attached_to
+          align_position = position_attached_element element
+        else
+          position_element element, current_position
+          align_position = current_position
+        end
```
Comment (jasonrclark) [inline shoes-core/lib/shoes/slot.rb]: Yeah, the lack of symmetry between attached/not there was bothering me. Will see what I can work up.

**Comment by jasonrclark** [comment]
So I think this should be ready to go (getting CI running while I do a quick scan through the full samples before bed here). Things I learned:

* `Shoes::Window` exists in Shoes 3 but doesn't appear to do anything except be a marker for saying things like `attach: Shoes::Window`. The class has no instance methods of its own, and greping the old code turned up minimal usage of it.
* Margins aren't respected in positioning of attached elements. This meant a change from `element_left` to `absolute_left` in my slot code. Fair enough.

**Comment by PragTob** [comment]
Ah you already fixed the one indentation thing.. all great then! :)

**Pull request merged**